libs/blocks/marketo-config/marketo-config.js

```diff
@@ -0,0 +1,259 @@
+  const { state, dispatch } = useContext(ConfiguratorContext);
+
+  const onChange = (prop, value) => {
+    dispatch({
+      type: 'SET_VALUE',
+      prop,
+      value,
+    });
+  };
+
+  return fieldsData.map((field) => {
+    const prop = field.prop;
+    const value = state?.[prop];
+    const required = field.required === 'yes';
+
+    if (!field.options) {
+      return html`
+        <${Input} label=${field.label} name=${prop} tooltip=${field.description} type="text" value=${value} onChange=${(value) => onChange(prop, value)} isRequired=${required} />
+      `;
+    }
+
+    let options = {};
+    try {
+      options = JSON.parse(field.options);
+    } catch {
+      const keyValuePairs = field.options.split(',').map((item) => {
+        const [key, val] = item.trim().split(':');
+        return key.trim() ? [key.trim(), `${(val || key).trim()} (${key.trim()})`] : ['', 'Choose an option...'];
+      });
+      options = Object.fromEntries(keyValuePairs);
+    }
+
+    if (typeof options !== 'object') {
+      options = { [field.options]: field.options };
+    }
+
+    return html`
+      <${Select} label=${field.label} name=${prop} options=${options} tooltip=${field.description} value=${value} onChange=${(value) => onChange(prop, value)} isRequired=${required} />
+    `;
+  });
+};
+
+const validateState = (state, panelsData) => {
+  let validatedState = {};
+
+  Object.values(panelsData).forEach(panelConfig => {
+    panelConfig.forEach(field => {
+      if (field?.prop) {
+        const key = field.prop;
+        if (key in state) {
+          validatedState[key] = state[key];
+        }
+      }
+    });
+  });
+
+  return validatedState;
+};
+
+const getPanels = (panelsData) => {
+  return Object.entries(panelsData).map(([panelName, panelConfig]) => ({
+    title: panelName.charAt(0).toUpperCase() + panelName.slice(1),
+    content: html`<${Fields} fieldsData=${panelConfig} />`,
+  }));
+};
+
+const Configurator = ({ title, blockClass, panelsData, lsKey }) => {
+  const context = useContext(ConfiguratorContext);
+  const { state } = context;
+
+  const panels = getPanels(panelsData);
+
+  const getUrl = () => {
+    const url = window.location.href.split('#')[0];
+    return `${url}#${utf8ToB64(JSON.stringify(state))}`;
+  };
+
+  const configFormValidation = () => {
+    const invalidInputs = document.querySelectorAll('.input-invalid');
+    const hasInputsValid = invalidInputs.length === 0;
+
+    invalidInputs.forEach((input) => {
+      const requiredPanel = input.closest('.accordion-item');
+      const requiredPanelExpandButton = requiredPanel.querySelector('button[aria-label=Expand]');
+      if (requiredPanelExpandButton) {
+        requiredPanelExpandButton.click();
+      }
+      input.focus();
+    });
+
+    return hasInputsValid;
+  };
+
+  useEffect(() => {
+    const validatedState = validateState(state, panelsData);
+    setPreferences(validatedState);
+    saveStateToLocalStorage(validatedState, lsKey);
```

Comment:
Can you elaborate? State should only change from a field update on user input, maybe we need to update formControl? 
https://preactjs.com/guide/v10/differences-to-react/#use-oninput-instead-of-onchange
I might need to dig a little more into what it means by "gets fired when an element's value is _committed_ by the user.".

Reply:
If you don't think there will be a ton of state updates, then it's probably fine. In general, if there are several successive state updates that overwrite each other, it can be more performant to add a debounce so you're updating the storage less often. If that's not a worry here, that's fine